### PR TITLE
api: remove `mapstructure` tags from`Port` struct

### LINKED ISSUE #11044: Go API: Port.Value does not populate when using event.Allocation()
### Nomad version
1.1.2

### Operating system and Environment details
Nomad running in Debian VM, go api running on MacOS

### Issue
The DynamicPorts Ports have all their "Value" fields turned into 0s when running event.Allocation()
This could be a simple mapstructure misconfiguration or something else.
https://github.com/hashicorp/nomad/blob/499fcebc42f733e7200df30ef9db76074b6c1bbc/api/resources.go#L101

### Reproduction steps
with Nomad go api, listen to event stream
With valid Allocation event (with Dynamic Ports visible through curl)
```
allocation, err := event.Allocation()
if err != nil {
        log.Fatal("Event allocation get failed")
}

log.Println(event.Payload["Allocation"]) // Sees the value
log.Println(allocation.Resources.Networks[0].DynamicPorts) // Value is now 0 in all ports
```

#### Expected Result
Value would be the same as the event Payload value

#### Actual Result
Value is always 0 for all Ports

### Job file (if appropriate)
N/A, all job files that require ports have this.

### PR DESCRIPTION
**Pull request by DerekStrickland**
This PR solves a defect in the deserialization of `api.Port` structs when returning structs from the`EventStream`. 

Previously, the `api.Port` struct's fields were decorated with both `mapstructure` and `hcl` tags to support the `network.port` stanza's use of the keyword `static` when posting a static port value. This works fine when posting a job and when retrieving any struct that has an embedded `api.Port` instance as long as the value is deserialized using `JSON` decoding. The `EventStream`, however, uses `mapstructure` to decode event payloads in the `api` package. `mapstructure` expects an underlying field named `static` which does not exist. The result was that the `Port.Value` field would always be set to `0`.

Upon further inspection, a few things became apparent.

- The struct already has `hcl` tags that support the indirection during job submission.
- Serialization/deserialization with both the `json` and `hcl` packages produce the desired result.
- The use of of the `mapstructure` tags provided no value as the `Port` struct contains only fields with primitive types.

This PR:
- Removes the `mapstructure` tags from the `api.Port` structs
- Updates the job parsing logic to use `hcl` instead of `mapstructure` when decoding `Port` instances.

Closes #11044

### CODE REVIEW

File: e2e/events/events_test.go [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+	"github.com/hashicorp/nomad/helper/uuid"
+	"github.com/hashicorp/nomad/nomad/structs"
+	"github.com/hashicorp/nomad/testutil"
+	"github.com/stretchr/testify/require"
+	"testing"
+)
+
+// TestEventStream_Ports tests that removing the mapstructure tags from the Port
+// struct results in Port values being serialized correctly to the event stream.
+func TestEventStream_Ports(t *testing.T) {
```
Comment (tgross) [inline e2e/events/events_test.go:17]: This test doesn't really cover the whole change, does it? It looks like we're testing that we can decode the serialized value for the ports, but not that we haven't changed the API. (Similar to my comment above for the `alloc_endpoint_test.go`)

Also, having a new nightly E2E test for this bug seems excessive; E2E tests really need to bring a lot of value to the party. Could we make these fields appear in the existing events E2E test?

File: command/agent/alloc_endpoint_test.go [hunk 2]
```diff
@@ -1121,3 +1121,29 @@ func TestHTTP_ReadWsHandshake(t *testing.T) {
 		})
 	}
 }
+
+// TestHTTP_AllocPort_Parsing tests that removing the mapstructure tags from the
+// Port struct has no adverse affects on serialization.
+func TestHTTP_AllocPort_Parsing(t *testing.T) {
```
Comment (tgross) [inline command/agent/alloc_endpoint_test.go:1127]: I think this test only verifies that the serialized object can still be round-tripped, not that we haven't changed the API? It's probably worth marking this test with a TODO to tear the test out once it's shipped, as it only verifies backcompat.

File: jobspec/parse_test.go [hunk 3]
```diff
@@ -1865,3 +1865,26 @@ func TestIncorrectKey(t *testing.T) {
 		t.Fatalf("Expected key error; got %v", err)
 	}
 }
+
+// TestPortParsing validates that the removal of the mapstructure tags on the Port
+// struct don't cause issues with HCL 1 parsing.
```
Comment (tgross) [inline jobspec/parse_test.go:1870]: This comment will be stale and mysterious after we ship this. Probably worth having a TODO here that points to a new issue examining whether we should have `mapstructure` tags on _any_ of the API at this point.

File: e2e/events/events_test.go (hunk 1 above)
Comment (DerekStrickland) [reply e2e/events/events_test.go:17]: oh yeah. I could probably find a way to cover this in an existing test.

File: jobspec/parse_test.go (hunk 3 above)
Comment (pkazmierczak) [reply jobspec/parse_test.go]: I added a comment that clarifies this.

File: e2e/events/events_test.go (hunk 1 above)
Comment (pkazmierczak) [reply e2e/events/events_test.go]: I added `network` stanzas into existing e2e job files.

File: command/agent/alloc_endpoint_test.go (hunk 2 above)
Comment (pkazmierczak) [reply command/agent/alloc_endpoint_test.go]: I think the [job_endpoint tests](https://github.com/hashicorp/nomad/blob/24e01efa843b4babe922cfc2007de58c514c8424/command/agent/job_endpoint_test.go#L2650) already check that we're not breaking the API.

Comment (tgross) [reply command/agent/alloc_endpoint_test.go]: Ok, and a quick smoke test confirms that:

```
$ nomad operator api "/v1/allocation/73105ebb-22a0-e7aa-621e-cbe72eb31220" | jq '.AllocatedResources.Shared.Networks'
[
  {
    "CIDR": "",
    "DNS": null,
    "Device": "",
    "DynamicPorts": [
      {
        "HostNetwork": "default",
        "Label": "www",
        "To": 8001,
        "Value": 31440
      }
    ],
    "IP": "127.0.0.1",
    "MBits": 0,
    "Mode": "bridge",
    "ReservedPorts": [
      {
        "HostNetwork": "private",
        "Label": "testing",
        "To": 8002,
        "Value": 9002
      }
    ]
  }
]
$ nomad operator api "/v1/allocation/73105ebb-22a0-e7aa-621e-cbe72eb31220" | jq '.SharedResources.Networks'
[
  {
    "CIDR": "",
    "DNS": null,
    "Device": "",
    "DynamicPorts": [
      {
        "HostNetwork": "default",
        "Label": "www",
        "To": 8001,
        "Value": 31440
      }
    ],
    "IP": "127.0.0.1",
    "MBits": 0,
    "Mode": "bridge",
    "ReservedPorts": [
      {
        "HostNetwork": "private",
        "Label": "testing",
        "To": 8002,
        "Value": 9002
      }
    ]
  }
]

```

**Review by tgross — APPROVED** [review]
This is looking good @pkazmierczak! There's some stray bits to clean up but once you've done that I think we're good to merge.

File: command/agent/event_endpoint_test.go [hunk 4]
```diff
@@ -202,3 +204,103 @@ func TestEventStream_QueryParse(t *testing.T) {
 		})
 	}
 }
+
+func TestHTTP_AllocPort_Parsing(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (pkazmierczak) [reply command/agent/event_endpoint_test.go:208]: good point 👍

File: command/agent/test-resources/parse-ports.nomad [hunk 5]
```diff
@@ -0,0 +1,32 @@
+job "parse-ports" {
```
[the remark this answers is not in the document]
Comment (pkazmierczak) [reply command/agent/test-resources/parse-ports.nomad:1]: right!

File: api/testutil.go [hunk 6]
```diff
@@ -0,0 +1,24 @@
+package api
+
+import (
+	"testing"
+
+	"github.com/hashicorp/nomad/api/internal/testutil"
+)
+
+// NewTestClient returns an API client useful for testing.
+func NewTestClient(t *testing.T) (*Client, error) {
```
[the remark this answers is not in the document]
Comment (pkazmierczak) [reply api/testutil.go:10]: Yup.

**Pull request merged**